crates/engine/tree/src/chain.rs

```diff
@@ -0,0 +1,142 @@
+use futures::Stream;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// The type that drives the chain forward.
+///
+/// A state machine that orchestrates the components responsible for advancing the chain
+///
+///
+/// ## Control flow
+///
+/// The [`ChainOrchestrator`] is responsible for controlling the pipeline sync and additional hooks.
+/// It polls the given `handler`, which is responsible for advancing the chain, how is up to the
+/// handler. However, due to database restrictions (e.g. exclusive write access), following
+/// invariants apply:
+///  - If the handler requests a pipeline run (e.g. [`PipelineAction::SyncPipeline`]), the handler
+///    must ensure that while the pipeline is running, no other write access is granted.
+///  - At any time the [`ChainOrchestrator`] can request exclusive write access to the database
+///    (e.g. if pruning is required), but will not do so until the handler has acknowledged the
+///    request for write access.
+///
+/// The [`ChainOrchestrator`] polls the [`ChainHandler`] to advance the chain and handles the
+/// emitted events. Requests and events are passed to the [`ChainHandler`] via
+/// [`ChainHandler::on_event`].
+#[must_use = "Stream does nothing unless polled"]
+pub struct ChainOrchestrator<T>
+where
+    T: ChainHandler,
+{
+    /// The handler for advancing the chain.
+    handler: T,
+    /// Controls pipeline sync.
+    pipeline: (),
+    /// Additional hooks (e.g. pruning) that can require exclusive access to the database.
+    hooks: (),
```

Comment:
I wonder if `Pruning` has been all along the wrong way to think about it, and instead we should have an `Archiver` hook, and we implement `Archiver` for a Static Files Hook and for a State Diffs hook?